### PR TITLE
[Pt1] Requirement-ize Tailoring, make Tailoring flexible enough to deal with new materials

### PR DESCRIPTION
**Pull request by dissociativity**
#### Summary
Balance "[Pt1] Requirement-ize Tailoring, make Tailoring flexible enough to deal with new materials"
#### Purpose of change
New materials are now scattered throughout the game for making clothing, also using requirements makes for less cluttered jsons

#### Describe the solution
add new requirement types, use them instead of adding lots of entries to each recipe
This is part one and mostly covers a finished torso.json and suits.json, the rest are partially done for early-game recipes so the player freezes less easily.

#### Describe alternatives you've considered
don't do this

#### Testing
load into game, take apart sheet, make recipes

### CODE REVIEW

**Comment by dissociativity** [comment]
I'll probably do more PRs as this happens, a better solution may be to obsolete "rag" for the newer materials and auto-replace the recipe stuff?

**Comment by cainiaowu** [comment]
you can make them into a "requirement" like 
https://github.com/CleverRaven/Cataclysm-DDA/blob/master/data/json/requirements/materials.json

**Comment by dissociativity** [comment]
neat, I'll work on that tomorrow

**Comment by dissociativity** [comment]
oops now the scope is huge, but I'll work on this for at least an hour a day and be done with it all without too much time, should make for far more tailoring flexibility in game, please let me know if specific items /should/ strictly be made from cotton that used rags before and I'll make them use a new cotton requirement instead of allowing non-stretchy fabrics in general.

**Comment by dissociativity** [comment]
wow, really need a proper json editor that'll point out syntax errors, the web linter just kind of cries at you and notepad++ just helps highlight syntax, not actually let me know where something is missing or broken

**Review by Maleclypse — DISMISSED** [review]
This is a pretty good update.  Thanks!

File: data/json/recipes/armor/torso.json [hunk 1]
```diff
@@ -1504,8 +1453,7 @@
     "time": "8 h 20 m",
     "reversible": true,
     "autolearn": true,
-    "using": [ [ "sewing_standard", 45 ], [ "fastener_large", 1 ] ],
-    "components": [ [ [ "fur", 30 ], [ "tanned_pelt", 5 ] ] ],
+    "using": [ [ "sewing_standard", 45 ], [ "fabric_fur_pelt", 5 ]  [ "fastener_large", 1 ] ],
```
Comment (Maleclypse) [inline data/json/recipes/armor/torso.json:1456]:
```suggestion
    "using": [ [ "sewing_standard", 45 ], [ "fabric_fur_pelt", 5 ],  [ "fastener_large", 1 ] ],
```

**Comment by Maleclypse** [comment]
I fixed the errors in the torso file but can't carry linting over in the edit function apparently.  But I think linting would be all you need to have this ready for merge.

**Comment by dissociativity** [comment]
> 
> 
> I fixed the errors in the torso file but can't carry linting over in the edit function apparently. But I think linting would be all you need to have this ready for merge.

the scope of this PR is actually to include all clothing files, but it can be reduced if you'd like the smaller changes in more immediately, suits.json and torso.json are both finished other than the lint, with a few minor changes to other jsons, mostly for early-game clothing to stop the player from freezing easily

**Comment by Maleclypse** [comment]
Smaller PRs that cover a coherent chapter of the entire work are best. It reduces the load on reviewers. And in the event that something slips through reviews and automated testing and bricks the build it’s easier to find the error.

**Comment by dissociativity** [comment]
part 1 is done and ready to merge I suppose then

File: data/json/recipes/armor/torso.json [hunk 2]
```diff
@@ -1601,8 +1539,7 @@
     "time": "1 h",
     "reversible": true,
     "autolearn": true,
-    "using": [ [ "sewing_standard", 4 ] ],
-    "components": [ [ [ "rag", 5 ] ] ]
+    "using": [ [ [ "sewing_standard", 4 ], [ "fabric_standard_nostretch", 5 ] ] ]
```
Comment (Maleclypse) [inline data/json/recipes/armor/torso.json:1542]:
```suggestion
    "using": [ [ "sewing_standard", 4 ], [ "fabric_standard_nostretch", 5 ] ]
```

**Comment by Maleclypse** [comment]
Are these working on your local machine?   `felt in fabric_felt is not a valid item template
(~[slow] ~[.])=> 03:14:30.091 ERROR : src/requirements.cpp:531 [void component::check_consistency(const string&) const] sheet_fur_patchwork in fabric_fur is not a valid item template`  I've got a ton of errors popping up here

**Comment by dissociativity** [comment]
> 
> 
> Are these working on your local machine? `felt in fabric_felt is not a valid item template (~[slow] ~[.])=> 03:14:30.091 ERROR : src/requirements.cpp:531 [void component::check_consistency(const string&) const] sheet_fur_patchwork in fabric_fur is not a valid item template` I've got a ton of errors popping up here

my bad, must've missed it and done some changes /after/ things were tested, and messed up a bit, I'll figure it out

**Comment by dissociativity** [comment]
this is odd,
`17:28:49.374 ERROR : src/requirements.cpp:531 [void component::check_consistency(const string&) const] felt in fabric_felt is not a valid item template
17:29:12.324 ERROR : src/requirements.cpp:531 [void component::check_consistency(const string&) const] sheet_fur_patchwork in fabric_fur is not a valid item template
17:29:20.616 ERROR : src/requirements.cpp:531 [void component::check_consistency(const string&) const] sheet_fur in fabric_fur is not a valid item template
17:29:21.585 ERROR : src/requirements.cpp:531 [void component::check_consistency(const string&) const] sheet_fur_patchwork in fabric_fur_pelt is not a valid item template
17:29:22.088 ERROR : src/requirements.cpp:531 [void component::check_consistency(const string&) const] sheet_fur in fabric_fur_pelt is not a valid item template`
all of these items as far as I'm aware exist in tailoring.json, why are they not considered valid as items within the materials.json?

**Comment by dissociativity** [comment]
oh, or not, must be dissociating a bit too much and missing a handful of things, where are the old "rag" and "leather" even stored? it's bizzare they're not in tailoring.json

**Comment by dissociativity** [comment]
okay, I think this is good to go?

**Comment by dissociativity** [comment]
wait, nope

**Comment by dissociativity** [comment]
nevermind, that was an issue on my end, it loads fine now, just doing a little testing

**Comment by dissociativity** [comment]
> 
> 
> The blueprint test failed because of a recipe you changed. You’d need to change what the faction camp recipe calls for to match your updated recipe.

not experienced with faction camp recipe stuff, how would I go about that?

**Comment by dissociativity** [comment]
okay, so I've compiled it all and it's still asking for the json formatter, what am I doing wrong? the exe for the json formatter is in the correct directory, does it need to be renamed?

**Comment by mqrause** [comment]
`Specified blueprint requirements of faction_base_modular_hub_core_kitchen_butchery_center does not match calculated requirements.` So it's `faction_base_modular_hub_core_kitchen_butchery_center` that needs updating. There might also be others, I didn't look through the whole report.
Also the issue isn't the formatting but that the requirements for the faction camp recipe(s) need to be adjusted since your changes affected some relevant recipes.

**Comment by dissociativity** [comment]
to which search responds "no items match your search" as there is no json named that at all, no-one has really taught me to use some sort of github search files for lines function yet, so my only choice is to go through them one by one and hope I bump into the recipe, and then do that again if the test fails again.

**Comment by dissociativity** [comment]
starting a new medication and having difficulty focusing is fun, let's get this PR finished

**Comment by dissociativity** [comment]
how should I resolve conflicts here? merge from upstream/master or?

**Comment by mqrause** [comment]
The recommended solution is to rebase onto master, but merging master into your branch works, too.

File: data/json/recipes/armor/torso.json [hunk 3]
```diff
@@ -891,8 +853,7 @@
     "time": "2 h",
     "reversible": true,
     "autolearn": true,
-    "using": [ [ "sewing_standard", 80 ] ],
-    "components": [ [ [ "felt_patch", 30 ] ] ]
+    "using": [ [ "sewing_standard", 80 ], [ "fabric_fur", 30 ] ]
```
Comment (Termineitor244) [inline data/json/recipes/armor/torso.json:856]:
```suggestion
    "using": [ [ "sewing_standard", 80 ], [ "fabric_felt", 30 ] ]
```

File: data/json/recipes/armor/torso.json [hunk 4]
```diff
@@ -1524,8 +1465,7 @@
     "time": "8 h 20 m",
     "reversible": true,
     "book_learn": [ [ "textbook_tailor", 3 ], [ "tailor_portfolio", 3 ], [ "recipe_fauxfur", 2 ] ],
-    "using": [ [ "sewing_standard", 45 ], [ "fastener_large", 1 ] ],
-    "components": [ [ [ "rag", 14 ] ], [ [ "faux_fur", 12 ] ] ],
+    "using": [ [ "sewing_standard", 45 ], [ "fabric_standard_nostretch", 14 ], [ "fabric_fur", 12 ], [ "fastener_large", 1 ] ],
```
Comment (Termineitor244) [inline data/json/recipes/armor/torso.json:1468]:
```suggestion
    "using": [ [ "sewing_standard", 45 ], [ "fabric_standard_nostretch", 14 ], [ "fabric_fauxfur", 12 ], [ "fastener_large", 1 ] ],
```

File: data/json/recipes/armor/suit.json [hunk 5]
```diff
@@ -141,9 +139,8 @@
     "type": "recipe",
     "copy-from": "armor_larmor",
     "time": "9 h",
-    "using": [ [ "sewing_standard", 76 ] ],
-    "qualities": [ { "id": "LEATHER_AWL", "level": 1 } ],
-    "components": [ [ [ "leather", 42 ], [ "tanned_hide", 7 ], [ "fur", 42 ], [ "tanned_pelt", 7 ] ] ]
+    "using": [ [ "sewing_standard", 76 ], [ "fabric_leather_fur_hide", 6 ] ],
```
Comment (Termineitor244) [inline data/json/recipes/armor/suit.json:142]:
```suggestion
    "using": [ [ "sewing_standard", 76 ], [ "fabric_leather_fur_hide", 7 ] ],
```

File: data/json/recipes/armor/suit.json [hunk 6]
```diff
@@ -238,18 +233,61 @@
...
+    "autolearn": true,
+    "using": [ [ "sewing_standard", 90 ], [ "fabric_standard_nostretch", 20 ] ],
+    "proficiencies": [ { "proficiency": "prof_closures" }, { "proficiency": "prof_closures_waterproofing" } ],
+    "components": [
+      [ [ "tunic", 1 ] ],
+      [ [ "shorts_cargo", 1 ], [ "shorts_denim", 1 ] ],
+      [ [ "vest", 1 ], [ "tacvest", 1 ] ],
+      [ [ "ragpouch", 2 ], [ "dump_pouch", 2 ], [ "fanny", 2 ] ],
+      [ [ "tool_belt", 1 ], [ "legrig", 1 ] ]
+    ]
```
Comment (Termineitor244) [inline data/json/recipes/armor/suit.json:290]: This was moved to another file and is now an exodii given recipe, see #51987
```suggestion
```

File: data/json/recipes/armor/suit.json [hunk 7]
```diff
@@ -735,7 +766,12 @@
     "skills_required": [ "fabrication", 6 ],
     "time": "12 h",
     "autolearn": true,
-    "using": [ [ "sewing_standard", 200 ] ],
+    "using": [
+      [ "sewing_standard", 200 ],
+      [ "fabric_standard_nostretch", 10 ],
+      [ "fabric_leather_hide", 4 ],
```
Comment (Termineitor244) [inline data/json/recipes/armor/suit.json:772]:
```suggestion
```

File: data/json/recipes/armor/suit.json [hunk 8]
```diff
@@ -991,8 +1021,7 @@
     "difficulty": 2,
     "time": "54 m",
     "book_learn": [ [ "textbook_tailor", 4 ], [ "manual_tailor", 5 ], [ "tailor_portfolio", 4 ] ],
-    "using": [ [ "sewing_standard", 40 ] ],
-    "components": [ [ [ "rag", 30 ] ] ]
+    "using": [ [ "sewing_standard", 40 ], [ "fabric_standard_nostretch", 20 ] ]
```
Comment (Termineitor244) [inline data/json/recipes/armor/suit.json:1024]:
```suggestion
    "using": [ [ "sewing_standard", 40 ], [ "fabric_standard_nostretch", 30 ] ]
```

File: data/json/recipes/armor/suit.json [hunk 9]
```diff
@@ -1255,10 +1282,9 @@
     "difficulty": 4,
     "time": "10 h",
     "book_learn": [ [ "textbook_tailor", 4 ], [ "manual_tailor", 5 ], [ "tailor_portfolio", 4 ] ],
-    "using": [ [ "sewing_standard", 50 ] ],
+    "using": [ [ "sewing_standard", 50 ], [ "fabric_standard_nostretch", 12 ], [ "fabric_neoprene", 8 ] ],
```
Comment (Termineitor244) [inline data/json/recipes/armor/suit.json:1285]:
```suggestion
    "using": [ [ "sewing_standard", 50 ], [ "fabric_standard_nostretch", 12 ], [ "fabric_neoprene", 12 ] ],
```

**Review by Termineitor244 — CHANGES_REQUESTED** [review]
I checked that all the numbers and requirements were in order with respect to their previous numbers and components, I found some that need to be adjusted (Like fax fur clothing using fur haha), but after this changes this should be ready to merge.

**Comment by dissociativity** [comment]
yeah, I dissociate a lot and make mistakes on larger stuff like this a lot oops

**Comment by Termineitor244** [comment]
You're welcome! Thanks for doing all this work! I don't know if it means anything, but I approve this changes.

**Review by Termineitor244 — APPROVED** [review]

**Comment by dissociativity** [comment]
this is ready to merge, just to be clear

**Comment by Termineitor244** [comment]
> this is ready to merge, just to be clear

New conflicts have arisen... I will comment about this PR in discord again, it would be really good to have it merged before conflicts arise even more.

File: data/json/recipes/armor/suit.json [hunk 10]
```diff
@@ -844,6 +844,7 @@
     "autolearn": true,
     "book_learn": [ [ "textbook_armwest", 3 ] ],
     "using": [ [ "sewing_standard", 30 ], [ "fabric_standard_nostretch", 26 ] ]
+    "components": [ [ [ "rag", 26 ] ], [ "fastener_large", 1 ] ]
```
Comment (Termineitor244) [inline data/json/recipes/armor/suit.json]:
```suggestion
    "using": [ [ "sewing_standard", 30 ], [ "fabric_standard_nostretch", 26 ], [ "fastener_large", 1 ] ]
    "components": [ [ [ "rag", 26 ] ] ]
```

**Review by Termineitor244 — CHANGES_REQUESTED** [review]

File: data/json/recipes/armor/suit.json [hunk 11]
```diff
@@ -857,8 +843,8 @@
     "time": "6 h",
     "autolearn": true,
     "book_learn": [ [ "textbook_armwest", 3 ] ],
-    "using": [ [ "sewing_standard", 35 ], [ "fastener_large", 1 ] ],
-    "components": [ [ [ "rag", 26 ] ] ]
+    "using": [ [ "sewing_standard", 30 ], [ "fabric_standard_nostretch", 26 ] ]
+    "components": [ [ [ "rag", 26 ] ], [ "fastener_large", 1 ] ]
```
Comment (Termineitor244) [inline data/json/recipes/armor/suit.json]:
```suggestion
    "using": [ [ "sewing_standard", 35 ], [ "fabric_standard_nostretch", 26 ], [ "fastener_large", 1 ] ]
```

File: data/json/recipes/armor/suit.json [hunk 12]
```diff
@@ -857,8 +843,8 @@
     "time": "6 h",
     "autolearn": true,
     "book_learn": [ [ "textbook_armwest", 3 ] ],
-    "using": [ [ "sewing_standard", 35 ], [ "fastener_large", 1 ] ],
-    "components": [ [ [ "rag", 26 ] ] ]
+    "using": [ [ "sewing_standard", 30 ], [ "fabric_standard_nostretch", 26 ] ]
```
Comment (Maleclypse) [inline data/json/recipes/armor/suit.json:846]:
```suggestion
    "using": [ [ "sewing_standard", 30 ], [ "fabric_standard_nostretch", 26 ] ],
```

File: data/json/recipes/armor/suit.json [hunk 13]
```diff
@@ -880,8 +866,8 @@
     "type": "recipe",
     "copy-from": "gambeson",
     "time": "6 h 45 m",
-    "using": [ [ "sewing_standard", 50 ], [ "fastener_large", 2 ] ],
-    "components": [ [ [ "rag", 52 ] ] ]
+    "using": [ [ "sewing_standard", 45 ], [ "fabric_standard_nostretch", 62 ] ]
+    "components": [ [ [ "fastener_large", 2 ] ] ]
```
Comment (Termineitor244) [inline data/json/recipes/armor/suit.json:870]:
```suggestion
    "using": [ [ "sewing_standard", 45 ], [ "fabric_standard_nostretch", 52 ], [ "fastener_large", 2 ] ]
```

**Review by Termineitor244 — CHANGES_REQUESTED** [review]

File: data/json/recipes/armor/suit.json (hunk 11 above)
Comment (Termineitor244) [reply data/json/recipes/armor/suit.json]: @Maleclypse I think this change was lost when you applied your changes... The fastener should be in "using" and the rags are already included in fabric_standard_nostretch.

File: data/json/recipes/armor/suit.json [hunk 14]
```diff
@@ -857,8 +843,8 @@
     "time": "6 h",
     "autolearn": true,
     "book_learn": [ [ "textbook_armwest", 3 ] ],
-    "using": [ [ "sewing_standard", 35 ], [ "fastener_large", 1 ] ],
-    "components": [ [ [ "rag", 26 ] ] ]
+    "using": [ [ "sewing_standard", 30 ], [ "fabric_standard_nostretch", 26 ] ],
+    "components": [ [ [ "rag", 26 ] ], [ "fastener_large", 1 ] ]
```
Comment (Maleclypse) [inline data/json/recipes/armor/suit.json:847]:
```suggestion
    "components": [ [ [ "fastener_large", 1 ] ] ]
```
```suggestion
    "components": [ [ [ "rag", 26 ] ], [ "fastener_large", 1 ] ]
```

File: data/json/recipes/armor/suit.json [hunk 15]
```diff
@@ -857,8 +843,8 @@
     "time": "6 h",
     "autolearn": true,
     "book_learn": [ [ "textbook_armwest", 3 ] ],
-    "using": [ [ "sewing_standard", 35 ], [ "fastener_large", 1 ] ],
-    "components": [ [ [ "rag", 26 ] ] ]
+    "using": [ [ "sewing_standard", 30 ], [ "fabric_standard_nostretch", 26 ] ],
+    "components": [ [ [ "fastener_large", 1 ] ] ]
```
Comment (Termineitor244) [inline data/json/recipes/armor/suit.json:847]:
```suggestion
    "using": [ [ "sewing_standard", 35 ], [ "fabric_standard_nostretch", 26 ], [ "fastener_large", 1 ] ]
```

**Review by Termineitor244 — CHANGES_REQUESTED** [review]

**Comment by Maleclypse** [comment]
Probably, between work and family distractions I probably should have let someone else fix conflicts today.

**Comment by Termineitor244** [comment]
> Probably, between work and family distractions I probably should have let someone else fix conflicts today.

Doesn't matter, thanks for realizing the changes!

**Pull request merged**